File: cpp/samples/query/localidquery.cpp
/**
 * @file localidquery.cpp
 * @brief Demonstrates the vista-sdk LocalIdQuery API
 * @details Shows how to build queries to match LocalId instances by primary/secondary items
 *          and metadata tags, using GmodPathQuery and MetadataTagsQuery as building blocks.
 */

#include <dnv/VistaSDK.h>

#include <iostream>
#include <string>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk LocalIdQuery Sample ===\n\n";

    {
        std::cout << "1. LocalIdQuery: Matching by primary item\n";
        std::cout << "--------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto version = vis.latest();
        const auto& gmod = vis.gmod(version);
        const auto& locations = vis.locations(version);

        // Build query for specific primary item (ignoring locations)
        auto primaryPath = GmodPath::fromShortPath("411.1/C101.31", gmod, locations);
        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-pressure");
        auto localId3 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if (primaryPath.has_value() && localId1.has_value() && localId2.has_value() && localId3.has_value())
        {
            auto pathQuery = GmodPathQueryBuilder::from(*primaryPath).withoutLocations().build();

            auto query = LocalIdQueryBuilder::create().withPrimaryItem(pathQuery).build();

            std::cout << "Query: Match LocalIds with primary item '411.1/C101.31' (any location)\n\n";

            std::cout << "Test LocalId 1: " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId1) << "\n\n";

            std::cout << "Test LocalId 2: " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId2) << "\n\n";

            std::cout << "Test LocalId 3: " << localId3->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId3) << " (different node)\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "2. LocalIdQuery: Combined query (primary item + tags)\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto version = vis.latest();
        const auto& gmod = vis.gmod(version);

        auto node411Opt = gmod.node("411.1");
        auto nodeC101Opt = gmod.node("C101");
        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure");
        auto localId3 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if (node411Opt.has_value() && nodeC101Opt.has_value() && localId1.has_value() && localId2.has_value() &&
            localId3.has_value())
        {
            const auto* node411 = *node411Opt;
            const auto* nodeC101 = *nodeC101Opt;

            // Build path query
            auto pathQuery = GmodPathQueryBuilder::create().withNode(*node411, true).withNode(*nodeC101, true).build();

            // Build tags query
            auto tagsQuery = MetadataTagsQueryBuilder::create().withTag(CodebookName::Quantity, "temperature").build();

            // Combine in LocalIdQuery
            auto query = LocalIdQueryBuilder::create().withPrimaryItem(pathQuery).withTags(tagsQuery).build();

            std::cout << "Query: Match LocalIds with primary='411.1/C101' AND qty-temperature\n\n";

            std::cout << "Test LocalId 1: " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId1) << " (has both)\n\n";

            std::cout << "Test LocalId 2: " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId2) << " (missing tag)\n\n";

            std::cout << "Test LocalId 3: " << localId3->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId3)
                      << " (C101.63 is a subnode of C101)\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "3. LocalIdQuery: Secondary item requirement\n";
        std::cout << "----------------------------------------------\n";

        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass");

        if (localId1.has_value() && localId2.has_value())
        {
            // Query requiring NO secondary item
            auto query1 = LocalIdQueryBuilder::create().withoutSecondaryItem().build();

            std::cout << "Query: Match LocalIds WITHOUT secondary item\n\n";
            std::cout << "Test LocalId 1 (no secondary): " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query1.match(*localId1) << "\n\n";

            std::cout << "Test LocalId 2 (has secondary): " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query1.match(*localId2) << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "4. Building queries from existing LocalIds\n";
        std::cout << "---------------------------------------------\n";

        // Parse a reference LocalId
        auto referenceLocalId =
            LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet");

        // Test against various LocalIds
        auto localId1 =
            LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet");
        auto localId2 =
            LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-temperature/cnt-exhaust.gas/pos-inlet");
        auto localId3 =
            LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure/cnt-exhaust.gas/pos-inlet");

        if (referenceLocalId.has_value() && localId1.has_value() && localId2.has_value() && localId3.has_value())
        {
            // Build query that matches this exact LocalId
            auto query = LocalIdQueryBuilder::from(*referenceLocalId).build();

            std::cout << "Reference LocalId: " << referenceLocalId->toString() << "\n\n";
            std::cout << "Query: Match exact copy of reference LocalId\n\n";

            std::cout << "Test LocalId 1 (exact match): \n  Matches: " << std::boolalpha << query.match(*localId1)
                      << "\n\n";
            std::cout << "Test LocalId 2 (different location): \n  Matches: " << std::boolalpha
                      << query.match(*localId2) << "\n\n";
            std::cout << "Test LocalId 3 (different tag): \n  Matches: " << std::boolalpha << query.match(*localId3)
                      << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "5. LocalIdQuery: withPrimaryItem using Nodes builder lambda\n";
        std::cout << "--------------------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto version = vis.latest();
        const auto& gmod = vis.gmod(version);

        auto node411Opt = gmod.node("411.1");
        auto nodeC101Opt = gmod.node("C101");
        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-pressure");
        auto localId3 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C102.31/meta/qty-temperature");

        if (node411Opt.has_value() && nodeC101Opt.has_value() && localId1.has_value() && localId2.has_value() &&
            localId3.has_value())
        {
            const auto* node411 = *node411Opt;
            const auto* nodeC101 = *nodeC101Opt;

            // Build base nodes query (can be reused)
            const auto baseNodes = GmodPathQueryBuilder::create().withNode(*node411, true).withNode(*nodeC101, true);

            // Build final query using const& copy (demonstrates reusability)
            auto query = LocalIdQueryBuilder::create().withPrimaryItem(baseNodes.build()).build();

            std::cout << "Query: Match LocalIds with primary containing nodes 411.1 AND C101 (any locations)\n\n";

            std::cout << "Test LocalId 1: " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId1) << "\n\n";

            std::cout << "Test LocalId 2: " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId2) << "\n\n";

            std::cout << "Test LocalId 3: " << localId3->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId3) << " (wrong second node)\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "6. LocalIdQuery: withPrimaryItem using base path + Path builder lambda\n";
        std::cout << "------------------------------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto version = vis.latest();
        const auto& gmod = vis.gmod(version);
        const auto& locations = vis.locations(version);

        auto basePath = GmodPath::fromShortPath("411.1/C101.31", gmod, locations);
        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-pressure");
        auto localId3 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if (basePath.has_value() && localId1.has_value() && localId2.has_value() && localId3.has_value())
        {
            auto query =
                LocalIdQueryBuilder::create()
                    .withPrimaryItem(
                        *basePath,
                        [](const GmodPathQueryBuilder::Path& builder) { return builder.withoutLocations().build(); })
                    .build();

            std::cout << "Query: Match LocalIds with primary '411.1/C101.31' (any location) using Path lambda\n\n";

            std::cout << "Test LocalId 1: " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId1) << "\n\n";

            std::cout << "Test LocalId 2: " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId2) << "\n\n";

            std::cout << "Test LocalId 3: " << localId3->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId3) << " (different node)\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "7. LocalIdQuery: withAnyNodeBefore usage\n";
        std::cout << "--------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto version = vis.latest();
        const auto& gmod = vis.gmod(version);
        const auto& locations = vis.locations(version);

        auto nodeC101Opt = gmod.node("C101");
        auto basePath = GmodPath::fromShortPath("411.1/C101.31", gmod, locations);
        auto localId1 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        auto localId2 = LocalId::fromString("/dnv-v2/vis-3-4a/411.1/E202.1/E31/meta/qty-pressure");
        auto localId3 = LocalId::fromString("/dnv-v2/vis-3-4a/621.21/S90/meta/qty-temperature");

        if (nodeC101Opt.has_value() && basePath.has_value() && localId1.has_value() && localId2.has_value() &&
            localId3.has_value())
        {
            const auto* nodeC101 = *nodeC101Opt;

            auto query =
                LocalIdQueryBuilder::create()
                    .withPrimaryItem(
                        *basePath,
                        [nodeC101](const GmodPathQueryBuilder::Path& builder) {
                            return builder
                                .withAnyNodeBefore(
                                    [nodeC101](
                                        const std::
                                            unordered_map<std::string, const GmodNode*, StringHash, std::equal_to<>>&)
                                        -> const GmodNode* { return nodeC101; })
                                .withoutLocations()
                                .build();
                        })
                    .build();

            std::cout << "Query: Match LocalIds with any node before C101 (ignoring nodes before C101)\n\n";

            std::cout << "Test LocalId 1 (411.1): " << localId1->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId1) << "\n\n";

            std::cout << "Test LocalId 2 (411.1/E202.1): " << localId2->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId2) << " (no C101)\n\n";

            std::cout << "Test LocalId 3 (621.21): " << localId3->toString() << "\n";
            std::cout << "  Matches: " << std::boolalpha << query.match(*localId3) << " (no C101)\n";
        }

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/query/local_id_query.rs
use vista_sdk::core::codebook_name::CodebookName;
use vista_sdk::core::gmod_path::OwnedGmodPath;
use vista_sdk::core::local_id::OwnedLocalId;
use vista_sdk::core::vis::Vis;
use vista_sdk::query::gmod_path_query_builder::OwnedGmodPathQueryBuilder;
use vista_sdk::query::local_id_query_builder::OwnedLocalIdQueryBuilder;
use vista_sdk::query::metadata_tags_query_builder::OwnedMetadataTagsQueryBuilder;

fn main() {
    println!("=== vista-sdk LocalIdQuery Sample ===\n");

    {
        println!("1. LocalIdQuery: Matching by primary item");
        println!("--------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis.gmod(version).expect("version should be valid");
        let locations = vis.locations(version).expect("version should be valid");

        let primary_path = OwnedGmodPath::from_short_path("411.1/C101.31", gmod, locations);
        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-pressure");
        let local_id3 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if let (Some(primary_path), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (primary_path, local_id1, local_id2, local_id3)
        {
            let path_query = OwnedGmodPathQueryBuilder::from_path(&primary_path)
                .expect("path builder should succeed")
                .without_locations()
                .expect("without_locations should succeed")
                .build();

            let query = OwnedLocalIdQueryBuilder::create()
                .with_primary_item_query(&path_query)
                .expect("with_primary_item_query should succeed")
                .build();

            println!("Query: Match LocalIds with primary item '411.1/C101.31' (any location)\n");

            println!("Test LocalId 1: {}", local_id1);
            println!("  Matches: {}\n", query.match_local_id(&local_id1));

            println!("Test LocalId 2: {}", local_id2);
            println!("  Matches: {}\n", query.match_local_id(&local_id2));

            println!("Test LocalId 3: {}", local_id3);
            println!(
                "  Matches: {} (different node)",
                query.match_local_id(&local_id3)
            );
        }

        println!();
    }

    {
        println!("2. LocalIdQuery: Combined query (primary item + tags)");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis.gmod(version).expect("version should be valid");

        let node411 = gmod.get_node("411.1");
        let node_c101 = gmod.get_node("C101");
        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure");
        let local_id3 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if let (Ok(node411), Ok(node_c101), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (node411, node_c101, local_id1, local_id2, local_id3)
        {
            let path_query = OwnedGmodPathQueryBuilder::create()
                .with_node_all_locations(node411, true)
                .expect("with_node_all_locations should succeed")
                .with_node_all_locations(node_c101, true)
                .expect("with_node_all_locations should succeed")
                .build();

            let tags_query = OwnedMetadataTagsQueryBuilder::create()
                .with_tag(CodebookName::Quantity, "temperature")
                .build();

            let query = OwnedLocalIdQueryBuilder::create()
                .with_primary_item_query(&path_query)
                .expect("with_primary_item_query should succeed")
                .with_tags(&tags_query)
                .expect("with_tags should succeed")
                .build();

            println!("Query: Match LocalIds with primary='411.1/C101' AND qty-temperature\n");

            println!("Test LocalId 1: {}", local_id1);
            println!(
                "  Matches: {} (has both)\n",
                query.match_local_id(&local_id1)
            );

            println!("Test LocalId 2: {}", local_id2);
            println!(
                "  Matches: {} (missing tag)\n",
                query.match_local_id(&local_id2)
            );

            println!("Test LocalId 3: {}", local_id3);
            println!(
                "  Matches: {} (C101.63 is a subnode of C101)",
                query.match_local_id(&local_id3)
            );
        }

        println!();
    }

    {
        println!("3. LocalIdQuery: Secondary item requirement");
        println!("----------------------------------------------");

        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/621.21/S90/sec/411.1/C101/meta/qty-mass");

        if let (Some(local_id1), Some(local_id2)) = (local_id1, local_id2) {
            let query = OwnedLocalIdQueryBuilder::create()
                .without_secondary_item()
                .expect("without_secondary_item should succeed")
                .build();

            println!("Query: Match LocalIds WITHOUT secondary item\n");

            println!("Test LocalId 1 (no secondary): {}", local_id1);
            println!("  Matches: {}\n", query.match_local_id(&local_id1));

            println!("Test LocalId 2 (has secondary): {}", local_id2);
            println!("  Matches: {}", query.match_local_id(&local_id2));
        }

        println!();
    }

    {
        println!("4. Building queries from existing LocalIds");
        println!("---------------------------------------------");

        let reference = OwnedLocalId::from_string(
            "/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet",
        );
        let local_id1 = OwnedLocalId::from_string(
            "/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet",
        );
        let local_id2 = OwnedLocalId::from_string(
            "/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-temperature/cnt-exhaust.gas/pos-inlet",
        );
        let local_id3 = OwnedLocalId::from_string(
            "/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-pressure/cnt-exhaust.gas/pos-inlet",
        );

        if let (Some(reference), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (reference, local_id1, local_id2, local_id3)
        {
            let query = OwnedLocalIdQueryBuilder::from_local_id(&reference)
                .expect("from_local_id should succeed")
                .build();

            println!("Reference LocalId: {}\n", reference);
            println!("Query: Match exact copy of reference LocalId\n");

            println!(
                "Test LocalId 1 (exact match):\n  Matches: {}\n",
                query.match_local_id(&local_id1)
            );
            println!(
                "Test LocalId 2 (different location):\n  Matches: {}\n",
                query.match_local_id(&local_id2)
            );
            println!(
                "Test LocalId 3 (different tag):\n  Matches: {}",
                query.match_local_id(&local_id3)
            );
        }

        println!();
    }

    {
        println!("5. LocalIdQuery: primary item via Nodes builder");
        println!("--------------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis.gmod(version).expect("version should be valid");

        let node411 = gmod.get_node("411.1");
        let node_c101 = gmod.get_node("C101");
        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-pressure");
        let local_id3 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C102.31/meta/qty-temperature");

        if let (Ok(node411), Ok(node_c101), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (node411, node_c101, local_id1, local_id2, local_id3)
        {
            let path_query = OwnedGmodPathQueryBuilder::create()
                .with_node_all_locations(node411, true)
                .expect("with_node_all_locations should succeed")
                .with_node_all_locations(node_c101, true)
                .expect("with_node_all_locations should succeed")
                .build();

            let query = OwnedLocalIdQueryBuilder::create()
                .with_primary_item_query(&path_query)
                .expect("with_primary_item_query should succeed")
                .build();

            println!(
                "Query: Match LocalIds with primary containing nodes 411.1 AND C101 (any locations)\n"
            );

            println!("Test LocalId 1: {}", local_id1);
            println!("  Matches: {}\n", query.match_local_id(&local_id1));

            println!("Test LocalId 2: {}", local_id2);
            println!("  Matches: {}\n", query.match_local_id(&local_id2));

            println!("Test LocalId 3: {}", local_id3);
            println!(
                "  Matches: {} (wrong second node)",
                query.match_local_id(&local_id3)
            );
        }

        println!();
    }

    {
        println!("6. LocalIdQuery: primary item via Path builder (without locations)");
        println!("---------------------------------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis.gmod(version).expect("version should be valid");
        let locations = vis.locations(version).expect("version should be valid");

        let base_path = OwnedGmodPath::from_short_path("411.1/C101.31", gmod, locations);
        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-5/meta/qty-pressure");
        let local_id3 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.63/meta/qty-temperature");

        if let (Some(base_path), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (base_path, local_id1, local_id2, local_id3)
        {
            let path_query = OwnedGmodPathQueryBuilder::from_path(&base_path)
                .expect("path builder should succeed")
                .without_locations()
                .expect("without_locations should succeed")
                .build();

            let query = OwnedLocalIdQueryBuilder::create()
                .with_primary_item_query(&path_query)
                .expect("with_primary_item_query should succeed")
                .build();

            println!("Query: Match LocalIds with primary '411.1/C101.31' (any location)\n");

            println!("Test LocalId 1: {}", local_id1);
            println!("  Matches: {}\n", query.match_local_id(&local_id1));

            println!("Test LocalId 2: {}", local_id2);
            println!("  Matches: {}\n", query.match_local_id(&local_id2));

            println!("Test LocalId 3: {}", local_id3);
            println!(
                "  Matches: {} (different node)",
                query.match_local_id(&local_id3)
            );
        }

        println!();
    }

    {
        println!("7. LocalIdQuery: withAnyNodeBefore usage");
        println!("--------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis.gmod(version).expect("version should be valid");
        let locations = vis.locations(version).expect("version should be valid");

        let base_path = OwnedGmodPath::from_short_path("411.1/C101.31", gmod, locations);
        let local_id1 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature");
        let local_id2 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/411.1/E202.1/E31/meta/qty-pressure");
        let local_id3 =
            OwnedLocalId::from_string("/dnv-v2/vis-3-4a/621.21/S90/meta/qty-temperature");

        if let (Some(base_path), Some(local_id1), Some(local_id2), Some(local_id3)) =
            (base_path, local_id1, local_id2, local_id3)
        {
            let path_query = OwnedGmodPathQueryBuilder::from_path(&base_path)
                .expect("path builder should succeed")
                .with_any_node_before("C101")
                .expect("C101 is in the path")
                .without_locations()
                .expect("without_locations should succeed")
                .build();

            let query = OwnedLocalIdQueryBuilder::create()
                .with_primary_item_query(&path_query)
                .expect("with_primary_item_query should succeed")
                .build();

            println!(
                "Query: Match LocalIds with any node before C101 (ignoring nodes before C101)\n"
            );

            println!("Test LocalId 1 (411.1): {}", local_id1);
            println!("  Matches: {}\n", query.match_local_id(&local_id1));

            println!("Test LocalId 2 (411.1/E202.1): {}", local_id2);
            println!(
                "  Matches: {} (no C101)\n",
                query.match_local_id(&local_id2)
            );

            println!("Test LocalId 3 (621.21): {}", local_id3);
            println!("  Matches: {} (no C101)", query.match_local_id(&local_id3));
        }

        println!();
    }
}
